pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -94,28 +97,328 @@ func (p *staticPolicy) Start(logger klog.Logger, s state.State) error {
+			blocks := []state.Block{}
+			for resourceName, req := range runnablePool {
+				if req > 0 {
+					blocks = append(blocks, state.Block{
+						NUMAAffinity: maskBits,
+						Size:         req,
+						Type:         resourceName,
+					})
+				}
+			}
+			exclusiveMemory[c.Name] = blocks
+			metrics.ResourceManagerContainerAssignments.WithLabelValues(metrics.ResourceManagerMemory, metrics.ResourceManagerSharedPod).Inc()
+			continue
+		}
+
+		// Allocate blocks
+		blocks := []state.Block{}
+		for resourceName, req := range reqs {
+			if req > 0 {
+				blocks = append(blocks, state.Block{
+					NUMAAffinity: maskBits,
+					Size:         req,
+					Type:         resourceName,
+				})
+			}
+		}
+		exclusiveMemory[c.Name] = blocks
+		metrics.ResourceManagerContainerAssignments.WithLabelValues(metrics.ResourceManagerMemory, metrics.ResourceManagerExclusivePod).Inc()
+
+		// If it's a restartable sidecar, its memory is permanently consumed from the pod's budget.
+		if podutil.IsRestartableInitContainer(&c) {
+			for resourceName, req := range reqs {
+				sidecarMemory[resourceName] += req
+			}
+		}
+	}
+
+	// Recalculate the pool available for app containers.
+	// This is the total pod budget minus what all sidecars are using.
+	appContainerPool := make(map[v1.ResourceName]uint64)
+	for resourceName, total := range podTotalMemory {
+		appContainerPool[resourceName] = total - sidecarMemory[resourceName]
+	}
+
+	// Second, iterate through regular containers, allocating from the remaining pool.
+	for _, c := range pod.Spec.Containers {
+		reqs, err := getContainerRequestedResources(logger, pod, &c)
+		if err != nil {
+			// This should not happen, as we have already validated the pod's resources.
+			return err
+		}
+
+		// A container with no memory requests will be part of the shared pool.
+		if len(reqs) == 0 {
+			continue
+		}
+		metrics.MemoryManagerPinningRequestTotal.Inc()
+
+		// Allocate blocks and consume from the pool
+		blocks := []state.Block{}
+		for resourceName, req := range reqs {
+			if req > 0 {
+				blocks = append(blocks, state.Block{
+					NUMAAffinity: maskBits,
+					Size:         req,
+					Type:         resourceName,
+				})
+				appContainerPool[resourceName] -= req
+			}
+		}
+		exclusiveMemory[c.Name] = blocks
+		metrics.ResourceManagerContainerAssignments.WithLabelValues(metrics.ResourceManagerMemory, metrics.ResourceManagerExclusivePod).Inc()
+	}
+
+	// The final shared pool is what remains after all exclusive allocations.
+	podSharedPoolBlocks := []state.Block{}
+	for resourceName, val := range appContainerPool {
+		if val > 0 {
+			podSharedPoolBlocks = append(podSharedPoolBlocks, state.Block{
+				NUMAAffinity: maskBits,
+				Size:         val,
+				Type:         resourceName,
+			})
+		}
+	}
+
+	// 5. Now that partitioning is successful, update the global machine state.
+	for resourceName, requestedSize := range podTotalMemory {
+		p.updateMachineState(machineState, maskBits, resourceName, requestedSize)
```

Comment:
I'm under the impression we are doing the updates wrong, which would lead to inconsistent internal state between memory blocks assignment and internal machine state. This will not probably affect allocations like I was suspecting, evne though I'm still checking, but can totally affect the kubelet restart, leading to startup failures. I don't have a failing test yet, but the idea is that after every test, `validateState` should pass - and it seems it isn't. I'll update as soon as I can give more concrete details.

Reply:
I added a new temp commit `temp-restore-state-flow` with the fix and the corresponding unit test for the restore state flow. The Memory manager `validateState` was counting multiple memory assignments, while the `AllocatePod` func was treating the whole pod-level resources allocation as a single assignment regardless of the containers. I updated the `validateState` func to also consider a single assignment if the pod has pod-level resources. Wdyt?